Throw error when adding private dicom tag without explicit VR

HwangTaehyun opened the pull request:
Fixes #1462.

Could you review this PR? @mrbean-bremen 
Thanks!

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [ ] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Throw error when adding private dicom tag without explicit VR #1463
- Check if the found entry from the registered dicom dictionary with UnknownTag instead of null

HwangTaehyun committed:
Throw error when adding private dicom tag without explicit VR

- Error message: "Unknown private tag <private creator> (xxxx,xxxx) has no VR defined"
- Add unit tests to DicomDatasetTest
  - Add_PrivateTagWithoutExplicitVR_ShouldThrow
  - Add_NonPrivateTagWithExplicitVR_ShouldBeAdded
  - AddOrUpdate_PrivateTagWithoutExplicitVR_ShouldThrow
  - AddOrUpdate_NonPrivateTagWithExplicitVR_ShouldBeAdded

Co-authored-by: Amdhj22 <amdhj22@gmail.com>

mrbean-bremen commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -1112,15 +1112,19 @@ private DicomDataset DoAdd(DicomItem item, bool allowUpdate)
-            if (entry == null)
-                throw new DicomDataException($"Tag {tag} not found in DICOM dictionary. Only dictionary tags may be added implicitly to the dataset.");
+            if (tag.IsPrivate) {
+                return DoAdd(DicomVR.LO, tag, values, allowUpdate);
```
Why would you always use LO for private tags? LO is for private creators, but private tags generally shall be set to UN if not known (e.g. not in the private dictionary). Maybe I don't understand something here?

HwangTaehyun replied:
When I AddOrUpdate PrivateDicomTag with a string value throws a validation error because this private tag's VR is regarded as AE and the fo-dicom validates the value with AE rules. Thus, I added this code to regard the private tag's VR as LO.

HwangTaehyun replied:
I think I paste the wrong link that explains private creators. I'll fix it.

HwangTaehyun replied:
```
- Private Data Element Tags' VR shall be LO if no explicit VR exists.
```

mrbean-bremen replied:
Have you seen that several tests fail due to this change? I still don't understand your fix: you cannot just assume that a private tag is of type LO, it may be of any type, and this would break any private tags that are of a type not fitting into LO. I haven't checked your issue yet (may have a look later), but without looking at it, I would guess that either the tag _should_ be AE because it is a known private tag, and just has the wrong value, or there is some bug determining the VR. Either way, we have to understand why the wrong VR is set first.

HwangTaehyun replied:
I added the tag with the following group number end element number.
```
ushort groupNumber = Convert.ToUInt16("00F1", 16);
ushort elementNumber = Convert.ToUInt16("1100", 16);
var dicomTag =                new DicomTag(groupNumber, elementNumber);
```

Is this tag the known private tag? What is the known private tag? Could you let me know that? Thank you!

mrbean-bremen replied:
To know that, you also need to know the private creator - a private tag is defined by the combination of private creator, group and element offset. What private creator did you set here?

HwangTaehyun replied:
I set private creator null or my company name "AIRS MEDICAL". However, both cases throw the same error! (regarded as AE value representation)

HwangTaehyun replied:
Where can I get that information of known private tags?

HwangTaehyun replied:
Thank you for reviewing this PR.

mrbean-bremen replied:
In the [private dictionary](https://github.com/fo-dicom/fo-dicom/blob/development/FO-DICOM.Core/Dictionaries/Private%20Dictionary.xml). You can also add your own private tags to the dictionary dynamically if needed.

HwangTaehyun replied:
Thank you! Then, below that code, it loads Private Dictionary or Dicom Dictionary according to private or not. Is it right?
https://github.com/fo-dicom/fo-dicom/blob/development/FO-DICOM.Core/DicomDataset.cs#:~:text=var%20entry%20%3D%20DicomDictionary.Default%5Btag.IsPrivate%20%3F%20GetPrivateTag(tag)%20%3A%20tag%5D%3B
However, I can't find my tag (00F1, 1100) in the [Private Dictionary](https://github.com/fo-dicom/fo-dicom/tree/development/FO-DICOM.Core/Dictionaries) you linked.

HwangTaehyun replied:
Also, when I debugged, I found out fo-dicom find my tag's VR only by comparing the value type of VR in the dictionary and the tag's value.
https://github.com/fo-dicom/fo-dicom/blob/development/FO-DICOM.Core/DicomDataset.cs#:~:text=/returns%3E-,private%20DicomDataset%20DoAdd%3CT%3E(DicomTag%20tag%2C%20IList,%7D,-///%20%3Csummary

```
private DicomDataset DoAdd<T>(DicomTag tag, IList<T> values, bool allowUpdate)
        {
            var entry = DicomDictionary.Default[tag.IsPrivate ? GetPrivateTag(tag) : tag];
            if (entry == null)
                throw new DicomDataException($"Tag {tag} not found in DICOM dictionary. Only dictionary tags may be added implicitly to the dataset.");

            DicomVR vr = null;
            if (values != null) vr = entry.ValueRepresentations.FirstOrDefault(x => x.ValueType == typeof(T));
            if (vr == null) vr = entry.ValueRepresentations.First();

            return DoAdd(vr, tag, values, allowUpdate);
        }
```

HwangTaehyun replied:
VR=AE has valueType which is System.String in the dicom dictionary , so my guid string tag value is regarded as AE.
![image](https://user-images.githubusercontent.com/42465090/196372183-5815ad83-3b2e-486e-abe9-0149834b40ea.png)
https://github.com/fo-dicom/fo-dicom/issues/1462

HwangTaehyun replied:
I think that Private Data Element Tags' VR shall be UN [you said](https://github.com/fo-dicom/fo-dicom/pull/1463#:~:text=set%20to%20UN%20if%20not%20known) if no explicit VR exists. Not LO

HwangTaehyun replied:
What do you think?

HwangTaehyun replied:
I wonder that dicom dictionary gets the valueRepresentations info where.

mrbean-bremen replied:
Sorry, I don't have time now, I may have another look tonight

HwangTaehyun replied:
Okay, Thank you!

mrbean-bremen commented:
I added some explanation for private tag handling to the question. I think that while this is not a bug, the code can be improved to handle this situation better. As I will be away for a couple of weeks, it will be a while until I can get to it though.

HwangTaehyun commented:
@mrbean-bremen Thank you, but I still think this is a bug When using AddOrUpdate Private Dicom Tag without explicit VR, the Private Dicom Tag's VR was inferred to AE VR in fo-dicom. 
fo-dicom's VR matching algorithm is not right to Private Dicom Tag.

mrbean-bremen commented:
> Thank you, but I still think this is a bug When using AddOrUpdate Private Dicom Tag without explicit VR, the Private Dicom Tag's VR was inferred to AE VR in fo-dicom.

I agree - this should be UN instead and I will have a look at this when I'm back. Nevertheless, the usage does not make sense if you don't define the VR yourself, so you have to adapt your code to define a VR for your private tags.

HwangTaehyun commented:
Yeah, if there is time for me, I want to fix this!

gofal commented:
Plese take care of the failing tests. When reading the discussion in the associated issue I have the impression that there is actually no need for a PR.

mrbean-bremen commented:
> Plese take care of the failing tests. When reading the discussion in the associated issue I have the impression that there is actually no need for a PR.

I still think that this is not a real bug, as creating a tag with an undefined VR is not a sensible use case. _If_ this is done nethertheless, it is unexpected that it is created as `AT` instead of `UN`. So while I would not see that as having any priority, but if somebody creates a working PR for this, that would be fine.

gofal commented:
> 

Agree. Or it could be even more strict. When adding a new value and the VR is not set explicitly as parameter and the VR also cannot be found in the DicomTagsDictionary, then an exception could be thrown instead of assuming an AE? UN would also be fine, but this is only a byte array internally, and in case of texts the serialization would be depending on Encoding used...

mrbean-bremen commented:
>  When adding a new value and the VR is not set explicitly as parameter and the VR also cannot be found in the DicomTagsDictionary, then an exception could be thrown instead of assuming an AE

I also thought about this, so maybe this is the way to go. It doesn't make much sense to add an unknown private tag without a VR given.
@HwangTaehyun - are you still working on this? If yes, what do you think about throwing an exception in this case?

HwangTaehyun commented:
@mrbean-bremen Yeah, I still do that. Then, Can I change the error message?
I'll change the error message from ```"Unhandled exception. FellowOakDicom.DicomValidationException: Content "f312e027-460d-4c8c-8d8a-807951b4241b" does not validate VR AE: value exceeds maximum length of 16 characters"``` to ```"Unhandled exception. FellowOakDicom.DicomValidationException: Content "f312e027-460d-4c8c-8d8a-807951b4241b" does not have explicit VR"```
If you have any other opinions, could you let me know?

mrbean-bremen commented:
I would phrase it a bit differently, something along the lines of: "Unknown private tag `<private creator>` (xxxx,xxxx) has no VR defined". 
Note that the exception should only be raised if creating a new tag manually - copying existing private tags with VR `UN` should be fine.

gofal commented:
and only changing the error message is of course not sufficient. The validation exception happens, after an value has been added and the vaildation fails. This exception, that we are talking here about, should be raised within the AddDicomTag method where no VR could be detected (neither as parameter nor from dictionary). 
As a unittest: if you add a private tag without explicit VR, but with value "ABC", then currently no exception will throw, because this value accidentally happens to match the AE validation. But in fact it should throw because a private tag needs an explicit VR.

HwangTaehyun commented:
I fixed my PR. Could you review that? @mrbean-bremen

mrbean-bremen commented on Contributors.md:
```diff
@@ -83,3 +83,5 @@
 * [Michael Ovens](https://github.com/MichaelOvens)
 * [Magnus Larsson](https://github.com/magla42)
 * [John Cupitt](https://github.com/jcupitt)
+* [Taehyun Hwang](https://github.com/HwangTaehyun)
+* [Hyojae Eum](https://github.com/Amdhj22)
```
Why there are two authors?

HwangTaehyun replied:
@Amdhj22 He is the coauthor for this PR!

mrbean-bremen commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -1112,15 +1112,22 @@ private DicomDataset DoAdd(DicomItem item, bool allowUpdate)
-            if (entry == null)
-                throw new DicomDataException($"Tag {tag} not found in DICOM dictionary. Only dictionary tags may be added implicitly to the dataset.");
+            if (tag.IsPrivate) {
+                string groupNumber = tag.Group.ToString("X4");
```
This will also fail for known private tags, where the VR is actually known.
Please also run the tests locally so you can see which ones fail (there are many tests failing now), and add a test for your changes.

HwangTaehyun replied:
How can I run the test? Could you let me know how to run the test script?

HwangTaehyun replied:
Is there any manual for running the test?

HwangTaehyun replied:
Oh, I see. I found that.
```
dotnet test ./Tests/FO-DICOM.Tests/FO-DICOM.Tests.csproj --configuration Release --framework net6.0 --blame --runtime win-x64 --logger:"trx;LogFileName=.\resultsnet60.xml" --collect:"XPlat Code Coverage" --settings coverlet.runsettings
```

HwangTaehyun replied:
I will add the unit test to my work!

HwangTaehyun replied:
I run the test on my m1 mac (osx.12-arm64). Once I run the test, many "dll not found exception" errors occurred.

HwangTaehyun replied:
```
Error Message:
   System.DllNotFoundException : Unable to load shared library 'Dicom.Native-macOS.dylib' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(libDicom.Native-macOS.dylib, 0x0001): tried: 'libDicom.Native-macOS.dylib' (no such file), '/usr/local/lib/libDicom.Native-macOS.dylib' (no such file), '/usr/lib/libDicom.Native-macOS.dylib' (no such file), '/Users/taehyun/github/HwangTaehyun/fo-dicom/Tests/FO-DICOM.Tests/bin/Release/net6.0/osx.12-arm64/libDicom.Native-macOS.dylib' (no such file), '/usr/local/lib/libDicom.Native-macOS.dylib' (no such file), '/usr/lib/libDicom.Native-macOS.dylib' (no such file)
  Stack Trace:
     at FellowOakDicom.Imaging.NativeCodec.Jpeg.JpegCodec.jpeg_std_error_8_MacOS(jpeg_error_mgr& err)
   at FellowOakDicom.Imaging.NativeCodec.Jpeg.JpegCodec.Decode(DicomPixelData oldPixelData, DicomPixelData newPixelData, DicomJpegParams jpegParams, Int32 frame)
   at FellowOakDicom.Imaging.NativeCodec.DicomJpegNativeCodec.Decode(DicomPixelData oldPixelData, DicomPixelData newPixelData, DicomCodecParams parameters)
```

HwangTaehyun replied:
If dll was not found on the host, then that was regarded as skip or fail?
```
Failed!  - Failed:    13, Passed:  1310, Skipped:    15, Total:  1338, Duration: 5 m 43 s - /Users/taehyun/github/HwangTaehyun/fo-dicom/Tests/FO-DICOM.Tests/bin/Release/net6.0/osx.12-arm64/fo-dicom.Tests.dll (net6.0)                                                                            /6m-5.1s
```

HwangTaehyun replied:
I think that was regarded as fail.

HwangTaehyun replied:
I would like to only run the test related to my work. Is it possible? (like DicomValidationTest.cs)

mrbean-bremen replied:
This should be possible, though I can't say how it's done on the Mac. Nethertheless, you still have to make sure all tests pass - currently your change breaks many of them.

HwangTaehyun replied:
I resolved it. I moved all tests to the parent of <kbd>FO-DICOM.Test</kbd> folder and only leave the DicomValidationTest.cs and run the test!

HwangTaehyun commented:
I fixed all failed test cases by adding a private tag with explicit VR and added the following test cases for my work.
Could you review this? @mrbean-bremen 

```
[Fact]
public void Add_PrivateTagWithoutExplicitVR_ShouldThrow()
{
    var dataset = new DicomDataset();

    var privateTag = new DicomTag(0x3001, 0x08, "PRIVATE");

    var e = Record.Exception(() => dataset.Add<string>(privateTag, "FO-DICOM"));
    Assert.IsType<DicomDataException>(e);
}

[Fact]
public void AddOrUpdate_PrivateTagWithoutExplicitVR_ShouldThrow()
{
    var dataset = new DicomDataset();

    var privateTag = new DicomTag(0x3001, 0x08, "PRIVATE");

    var e = Record.Exception(() => dataset.AddOrUpdate<string>(privateTag, "FO-DICOM"));
    Assert.IsType<DicomDataException>(e);
}
```

mrbean-bremen commented on Tests/FO-DICOM.Tests/DicomDatasetTest.cs:
```diff
@@ -604,6 +604,28 @@ public void AddOrUpdate_DicomItemOnExistingPrivateTag_PrivateGroupShouldCorrespo
+
+            var e = Record.Exception(() => dataset.AddOrUpdate<string>(privateTag, "FO-DICOM"));
+            Assert.IsType<DicomDataException>(e);
+        }
```
These tests look good - what is missing is the reverse test, e.g. make sure it does not throw if a VR is given (though that is already there in the other tests, so not really important), but more important, that it does not throw if you add a known private tag without a VR (e.g. where the definition is loaded from `Private Dictionary.xml`).

HwangTaehyun replied:
Ok.

And I also wonder what does "xx" mean? Can "xx" be all hex number? If not, does it mean real "xx" string?
```
The recommended convention for referencing a Private Data Element is (gggg,xxee,"pcde"), where gggg is the group number, xx is the string “xx”, ee is the element number within a reserved block, and pcde is the quoted value of the Private Creator Data Element that reserved the block, e.g., (0029,xx43,"Acme_CT_Parameters"). Alternatively, when a block of Private Data Elements is being described, one may factor out the description of the Private Creator Data Element value, e.g., Private Creator Data Element (0029,00xx) = "Acme_CT_Parameters", and (0029,xx43), (0029,xx44), etc.
```
```
<tag group="0029" element="xx80" vr="IS" vm="1">View Center</tag>
<tag group="0029" element="xx81" vr="IS" vm="1">View Size</tag>
<tag group="0029" element="xx82" vr="IS" vm="1">View Zoom</tag>
<tag group="0029" element="xx83" vr="IS" vm="1">View Transform</tag>
<tag group="6001" element="xx00" vr="LT" vm="1">Unknown</tag>
```

HwangTaehyun replied:
And I can't find any logic to match a known private tags' original VR.
I think DoAdd always set VR by matching the value type.
https://github.com/fo-dicom/fo-dicom/blob/development/FO-DICOM.Core/DicomDataset.cs#:~:text=instance.%3C/returns%3E-,private%20DicomDataset%20DoAdd%3CT%3E(DicomTag%20tag%2C%20IList%3CT%3E%20values%2C%20bool%20allowUpdate),-%7B

I think if a private tag is known, we should add this to the dataset directly with known private tags' VR without more checks. However, now fo-dicom always set VR by matching value type and use registered VR if VR is null.

Can you tell me if there's something I missed?

HwangTaehyun replied:
@mrbean-bremen I added reverse test you said and fix other test issues! I have only known private tag problem!

mrbean-bremen replied:
> And I also wonder what does "xx" mean? Can "xx" be all hex number? If not, does it mean real "xx" string?

As said in your quote, it is meant as an xx string, just as a convention for representing private tags. This byte only depends on the location of the private creator and is not part of the private tag definition, thus the placeholder "xx".

> And I can't find any logic to match a known private tags' original VR.

I only know that both the standard and private dictionaries are read from the xml files at startup, but I have to check myself how the matching is done (I expect that it is done, otherwise I would consider it a bug). Probably will have a look tonight, don't have the time right now...

HwangTaehyun replied:
I really thank you for your answer! I got it!

mrbean-bremen replied:
Here is the kind of test I meant that should pass:
```cs
    [Fact]
    public void AddOrUpdate_KnownPrivateTag()
    {
        var dataset = new DicomDataset();

        // add some tag that is defined in the private dictionary
        var privateTag = new DicomTag(0x0023, 0x17, "AMICAS0");
        // this should not throw 
        ds.Add(privateTag, 42u);
        // the tag shall have the correct VR
        Assert.Equal(DicomVR.UL, ds.GetDicomItem<DicomItem>(privateTag).ValueRepresentation);
    }
```

HwangTaehyun replied:
Yeah, Thank you. And I finally found the reason.
We should compare found entry with DicomDictionary.UnknownTag instead of null. DicomDictionary return DicomDictionary.UnknownTag if they can't find the entry from the registered dicom dictionary.
https://github.com/fo-dicom/fo-dicom/blob/development/FO-DICOM.Core/DicomDictionary.cs#:~:text=return-,UnknownTag,-%3B

UnknownTag entry passes all null checks and call DoAdd function. So, UnknownTag has many VRs, I understand why it throws like error VR: AE validation matching for the unregistered private tag!

HwangTaehyun replied:
@mrbean-bremen I fixed and pushed a new commit. Could you check that?

HwangTaehyun committed:
Check if the found entry from the registered dicom dictionary with UnknownTag instead of null

- Fix Add_RegularTags_ShouldBeSortedInGroupElementOrder test to use registered tags

mrbean-bremen reviewed (approved):
Thank, that looks good to me now. @gofal may have another look...

HwangTaehyun commented:
@gofal Could you review this PR? Thank you!

gofal commented on ChangeLog.md:
```diff
@@ -12,11 +12,12 @@
 * Added support for DICOM supplement 225, Multi-Fragment video transfer syntax (#1469)
 * Added property to omit adding the default Implicit VR Little Endian transfer syntax for CStoreRequest (#1475)
 * Fix blanking of ValueElements in the anonymizer (#1491)
+* Throw error when adding private dicom tag without explicit VR
```
This is a breaking change, because now the library throws where there has not been an exception before.

HwangTaehyun replied:
@gofal Yeah, I think so. I will reflect it!

gofal commented on Tests/FO-DICOM.Tests/Imaging/DicomOverlayDataTest.cs:
```diff
@@ -27,8 +27,8 @@ public void Constructor_GroupSpecified_GroupTransferredToDescription()
-            const string expected = "Subtype 6005";
-            const ushort group = 0x6005;
+            const string expected = "Subtype 6006";
+            const ushort group = 0x6006;
```
this overlays are private groups. And private groups are meant to be of odd number. You should not have to change the group number.
If this test fails, then some methods within DicomOverlayData should be updated to add VR

HwangTaehyun replied:
I think that was not a private tag cause Private Dictionary does not include them. I roll back all DicomOverlayDataTest.cs again.

HwangTaehyun replied:
@gofal I fixed it and there is no problem with the tests! Could you review it again?

mrbean-bremen commented on Tests/FO-DICOM.Tests/Imaging/DicomOverlayDataTest.cs:
```diff
@@ -17,10 +17,10 @@ public void Constructor_GroupSpecified_GroupTransferredToDescription()
             const ushort group = 0x6002;
 
-            var dataset = new DicomDataset();
+            DicomDataset dataset = new DicomDataset();
```
Why did you change all these `var` to the actual type? I see no reason for this.

HwangTaehyun replied:
My lint works. However, I think the explicit type is more reasonable. Would it be better to change it back?

HwangTaehyun replied:
@mrbean-bremen I roll back these changes. It does not need to be changed this time!

mrbean-bremen replied:
Strange - I would expect that cases like `DicomDataset dataset = new DicomDataset();` would get a remark from the linter to better use `var`, because this is redundant. At least Resharper does this, I think.

gofal replied:
some time ago I had the ping-pong where some contributors commited spaces and with the next PR they where changed to tabs and then the next PR changed back to spaces, etc.. 😄 

Therefore there is a .editorconfig file to force same styling.
There is already that var-handling:
https://github.com/fo-dicom/fo-dicom/blob/c90e4b059bf81dc3a58b4bc1f6f2bb511c33f5bd/FO-DICOM.Core/.editorconfig#L71
obviously your linter does not read this editorconfig file 😞

HwangTaehyun replied:
@gofal Yeah, but [Tests/FO-DICOM.Tests/Imaging/DicomOverlayDataTest.cs](https://github.com/fo-dicom/fo-dicom/pull/1463/files#diff-6c64f40ffc924b4eed0afcb3123d46439690d5d2537b2c348f3da07a1a0ea098) is not under FO-DICOM.Core/.editorconfig. Thus, my linter does not work for the Test files.
How about moving <kbd>.editorconfig</kbd> to the fo-dicom project folder?

HwangTaehyun replied:
Though, my DicomDataset.cs under FO-DICOM.Core is not applied by .editorconfig like you said.

HwangTaehyun replied:
@gofal Could you let me know where should I fix more?

HwangTaehyun replied:
@gofal Can I get some feedback?

gofal replied:
thanks, it's fine

HwangTaehyun replied:
@gofal Thank you!
